phoenix-core/src/main/java/org/apache/phoenix/util/ScanUtil.java

```diff
@@ -953,4 +953,80 @@ public static int getClientVersion(Scan scan) {
     }
+
+    /**
+     * Split the key ranges into multiple groups along the given boundaries
+     *
+     * @param boundaries
+     *     The boundaries is a byte[] list like "b0, b1, ..., bn" which forms
+     *     space (UNBOUND, b0), [b0, b1), ..., [bn, UNBOUND)
+     * @param keyRanges
+     *     The key ranges to split along the given boundaries. Coalesced.
+     * @return
+     *     key ranges in two dimensions.
+     */
+    public static List<List<KeyRange>> splitKeyRangesByBoundaries(List<byte[]> boundaries, List<KeyRange> keyRanges) {
+        if (boundaries == null || keyRanges == null || keyRanges.size() == 0) {
+            return null;
+        }
+
+        ScanUtil.BytesComparator comparator = ScanUtil.getComparator(true, SortOrder.ASC);
+        int countOfBoundaries = boundaries.size();
+        int countOfKeyRanges = keyRanges.size();
+        int indexOfBoundary = 0;
+        int indexOfKeyRange = 0;
+
+        List<List<KeyRange>> keyRangesGrouped = Lists.newArrayListWithExpectedSize(1);
+        List<KeyRange> keyRangesOfGroup = Lists.<KeyRange>newArrayList();
+        keyRangesGrouped.add(keyRangesOfGroup);
+
+        KeyRange currentKeyRange = null;
+        while (indexOfKeyRange < countOfKeyRanges) {
+            if (currentKeyRange == null) {
+                currentKeyRange = keyRanges.get(indexOfKeyRange);
+            }
+
+            if (indexOfBoundary >= countOfBoundaries) {
+                keyRangesOfGroup.add(currentKeyRange);
+                currentKeyRange = null;
+                indexOfKeyRange++;
+                continue;
+            }
+
+            byte[] regionBoundary = boundaries.get(indexOfBoundary);
+            int lowerToUpper = currentKeyRange.compareLowerToUpperBound(
+                    regionBoundary, 0, regionBoundary.length, false, comparator);
+            if (lowerToUpper < 0) {
+                int upperToUpper = currentKeyRange.compareUpperRange(regionBoundary, 0, regionBoundary.length, false);
+                if (upperToUpper >= 0) {
+                    keyRangesOfGroup.add(KeyRange.getKeyRange(
+                            currentKeyRange.getLowerRange(), currentKeyRange.isLowerInclusive(), regionBoundary, false));
+                    if (upperToUpper > 0) {
+                        currentKeyRange = KeyRange.getKeyRange(regionBoundary, true,
+                                currentKeyRange.getUpperRange(), currentKeyRange.isUpperInclusive());
+                    } else {
+                        currentKeyRange = null;
+                        indexOfKeyRange++;
+                    }
+                } else {
+                    keyRangesOfGroup.add(currentKeyRange);
+                    currentKeyRange = null;
+                    indexOfKeyRange++;
+                    continue;
+                }
+            }
+
+            // Increase region boundary
+            indexOfBoundary++;
+            keyRangesOfGroup = Lists.<KeyRange>newArrayList();
```

Comment:
Regarding the first comment, the current design is that, before the last region which has non-empty query key range list, for every region which has no query key range, we generate empty query key range list for it.

Regarding the second comment, could you show me more compact and cleaner code without using continue? I'm very curious how it would look like

Reply:
I changed to return List<Pair<Integer, List>>, your first concert should have gone.